### ISSUE TITLE
Energy derivatives with TorchForce

### ISSUE THREAD

**Issue by jharrymoore** [openmm/openmm#4516]
Hi,

I have a system where I define a `CustomCVForce` which containst several `TorchForce` objects with the energy being a linear combination, controlled by a global parameter.  The `TorchForce` objects also have a dependence on  the same global parameter - could I check whether, if I were to compute the derivative of the CV w.r.t. the global parameter, this is just returning the derivative of the energy function, not accounting for the extra dependence on the global parameter of the `TorchForce` objects in the expression?

**Comment by peastman** [comment]
It should be returning the total derivative.  It first evaluates the derivative for the contained force:

https://github.com/openmm/openmm/blob/2d4372d7ba8de7e02b4af9cb57f4ef29cd146f34/platforms/common/src/CommonKernels.cpp#L5582

and then multiplies by the derivative from the expression:

https://github.com/openmm/openmm/blob/2d4372d7ba8de7e02b4af9cb57f4ef29cd146f34/platforms/common/src/CommonKernels.cpp#L5608-L5610

**Comment by jharrymoore** [comment]
Thanks Peter, so this would look like adding the following method to TorchForce 
```
void TorchForce::addEnergyParameterDerivative(const string &name) {
  for (int i = 0; i < globalNames.size(); i++) {
    if (name == globalNames[i].name) {
      energyParameterDerivatives.push_back(i);
    }
  }
}
```
and then presumably in the platform-specific code within TorchForce the derivatives w.r.t the specified global parameters need to be calculated and pushed to the context when `calcForcesAndEnergy()` is called.  If I have the right approach, could you point me to somewhere in the code where this is done? Many thanks

**Comment by peastman** [comment]
Exactly.  Here's an example of code that accumulates parameter derivatives for the reference platform:

https://github.com/openmm/openmm/blob/2d4372d7ba8de7e02b4af9cb57f4ef29cd146f34/platforms/reference/src/ReferenceKernels.cpp#L471-L473

You just add the derivative to the appropriate entry in the map.  For CUDA and OpenCL it's a little more complicated, because you first have to tell it which parameters you'll be computing derivatives for:

https://github.com/openmm/openmm/blob/2d4372d7ba8de7e02b4af9cb57f4ef29cd146f34/platforms/common/src/CommonKernels.cpp#L5464-L5468

When you execute the force, you record them similarly to the reference platform.

https://github.com/openmm/openmm/blob/2d4372d7ba8de7e02b4af9cb57f4ef29cd146f34/platforms/common/src/CommonKernels.cpp#L5604-L5606

**Comment by peastman** [comment]
Parameter derivatives for TorchForce were added in https://github.com/openmm/openmm-torch/pull/143.
